ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala

```diff
@@ -0,0 +1,156 @@
+        intercept[UnorderedException[Int]] {
+          buffer.push(13, 2)
+        }.getMessage shouldBe s"Elements appended to the buffer should have strictly increasing offsets: 13 vs 13"
+      }
+    }
+
+    "range end with equal offset added" should {
+      "accept it" in withBuffer(3) { buffer =>
+        buffer.push(13, Int.MaxValue)
+        buffer.slice(0, 13) shouldBe Vector(8 -> 16, 13 -> 26)
+      }
+    }
+
+    "range end with greater offset added" should {
+      "not allow new element with lower offset" in withBuffer(3) { buffer =>
+        buffer.push(15, Int.MaxValue)
+        buffer.slice(0, 13) shouldBe Vector(8 -> 16, 13 -> 26)
+        intercept[UnorderedException[Int]] {
+          buffer.push(14, 28)
+        }.getMessage shouldBe s"Elements appended to the buffer should have strictly increasing offsets: 15 vs 14"
+      }
+    }
+  }
+
+  "getEvents" when {
+    "called with valid range" should {
+      val bufferElements = increasinglySparse(10).map(offset => offset -> offset * 2)
+
+      "return the correct series" in withBuffer(10, bufferElements) { buffer =>
+        val rangesGen = for {
+          startExclusive <- Gen.choose(0, bufferElements.last._1)
+          endInclusive <- Gen.choose(startExclusive, bufferElements.last._1)
+        } yield (startExclusive, endInclusive)
+
+        forAll(rangesGen, minSuccessful(100)) { case (startExclusive, endInclusive) =>
+          buffer.slice(startExclusive, endInclusive) shouldBe bufferElements
+            .dropWhile(_._1 <= startExclusive)
+            .takeWhile(_._1 <= endInclusive)
+        }
+      }
+    }
+
+    "called with endInclusive exceeding buffer range" should {
+      "fail with exception" in withBuffer() { buffer =>
+        intercept[RequestOffBufferBounds[Int]] {
+          buffer.slice(4, 15) shouldBe Vector(5 -> 10, 8 -> 16, 13 -> 26)
+        }.getMessage shouldBe s"Request endInclusive (15) is higher than bufferEnd (13)"
+      }
+    }
+
+    "called after push" should {
+      "always see the most recent updates" in withBuffer(1000, Vector.empty) { buffer =>
+        (0 until 1000).foreach(idx => buffer.push(idx, idx)) // fill buffer to max size
+
+        (0 until 10000).foldLeft[Assertion](Succeeded) {
+          case (failed: Failed, _) => failed
+          case (_, idx) =>
+            val expected = ((idx + 1) to (100 + idx)).map(idx => idx -> idx)
+            buffer.push(1000 + idx, 1000 + idx)
+            buffer.slice(idx, 100 + idx) should contain theSameElementsInOrderAs expected
+        }
+      }
+    }
+  }
+
+  "prune" when {
+    "element found" should {
+      "prune inclusive" in withBuffer() { buffer =>
+        buffer.prune(5)
+        buffer.slice(0, LastOffset) shouldBe Vector(8 -> 16, 13 -> 26)
+      }
+    }
+
+    "element inserted" should {
+      "prune inclusive" in withBuffer() { buffer =>
+        buffer.prune(6)
+        buffer.slice(0, LastOffset) shouldBe Vector(8 -> 16, 13 -> 26)
+      }
+    }
```

Comment:
Yes, it is basically asserting the end inclusive behavior: whether it lands on a vacant offset or not

Reply:
How about changing `element inserted` to `element not present`? I find it a bit counter-intuitive because after reading the test case header I went for looking where element `6` was inserted :)